### ISSUE TITLE
Questions

### ISSUE THREAD

**Issue by mavidser** [FlashforgeOfficial/AD5M_Series_Klipper#1]
This repo seems to be a slightly modified klipper 0.11 release. Can we re-push this with new commits on top, so maintainability with future klipper releases gets easier? It would allow to pull in upstream changes, and give visibility into changes from upstream that FlashForge did.

I [ran a diff with the v0.11.0 release](https://gist.github.com/mavidser/33ac2492c6efcf0dac4e217f3c9cb812) and copied over changes to this repository onto [this branch](https://github.com/mavidser/klipper/tree/ff-a5m).  Feel free to take that as a base, or recreate your own.

[This](https://github.com/mavidser/klipper/commit/3373ea22d13665e51b23eddb25191af42e855814) is the bulk of the changes (which mostly look like debugging), The other changes are :
- Deleted`.github` : Makes sense
- Deleted .gitignore : Not needed
- Present .'pyc' files, due to the removed .gitignore
- The generated `.bin` and `.hex` files. 
- Updated README
- Deleted Makefile : I do not understand why. This makes it harder to trust binaries.

Also, specific questions regarding this repo:
- Will you upload the .cfg file used for the 5M? This doesn't seem any useful without it.
- Why has the Makefile been removed? Can we get the build instructions please.
- What are the steps to flash this firmware?

**Comment by jlodew0** [comment]
I second that.

**Comment by mavidser** [comment]
Yes, looks like they might not do much for this. Come on @FlashforgeOfficial , why lock down good hardware from people who actually want to use it?

Good news is their config files are present in the [2.3.7 firmware](https://www.flashforge.com/download-center/144). 
[The config file](https://gist.github.com/mavidser/1af1a46c339a9e5733b14e95c667847d)

**Comment by mavidser** [comment]
Also, it could be that FF doesn't intends to open up more functionality and this repo is very likely only for GPL compliance. 

Luckily its almost stock Klipper (and we might be able to build and replace the klipper binaries with newer versions in their latest provided firmware).

There also seems to be active work in the community to set up external Mainsail+Moonraker.

https://github.com/g992/flashforge-ad5m-5mpro-research/issues/6

**Comment by inkeliz** [comment]
I'm not a lawyer, but if it don't have any proper way to build and no way to flash: is impossible to verify if the content of this repo is the actual firmware used by Flashforge.

I think that repo is only one minimal-effort to comply with the license. But, in the current state I'm almost sure that it still not complaint. Have one random "pre-compiled binary" is another proof of that. The lack of any scripts to build immediately violates the definition of "Corresponding Source", far I understand.

**Comment by consp** [comment]
> Luckily its almost stock Klipper

It's a direct copy of the v0.11.0 tag. I checked that out and compared the klippy and src dirs:

```
$ diff -r klipper/src AD5M_Series_Klipper/src
```
no changes there

```
diff -r klipper/klippy AD5M_Series_Klipper/klippy
Only in AD5M_Series_Klipper/klippy/chelper: c_helper.so
diff -r klipper/klippy/chelper/__init__.py AD5M_Series_Klipper/klippy/chelper/__init__.py
269,275c269,275
<         if check_build_code(srcfiles+ofiles+[__file__], destlib):
<             if check_gcc_option(SSE_FLAGS):
<                 cmd = "%s %s %s" % (GCC_CMD, SSE_FLAGS, COMPILE_ARGS)
<             else:
<                 cmd = "%s %s" % (GCC_CMD, COMPILE_ARGS)
<             logging.info("Building C code module %s", DEST_LIB)
<             do_build_code(cmd % (destlib, ' '.join(srcfiles)))
---
>        # if check_build_code(srcfiles+ofiles+[__file__], destlib):
>        #     if check_gcc_option(SSE_FLAGS):
>        #         cmd = "%s %s %s" % (GCC_CMD, SSE_FLAGS, COMPILE_ARGS)
>        #     else:
>        #         cmd = "%s %s" % (GCC_CMD, COMPILE_ARGS)
>        #     logging.info("Building C code module %s", DEST_LIB)
>        #     do_build_code(cmd % (destlib, ' '.join(srcfiles)))
Only in AD5M_Series_Klipper/klippy/chelper: __pycache__
diff -r klipper/klippy/chelper/stepcompress.c AD5M_Series_Klipper/klippy/chelper/stepcompress.c
29c29
< #define QUEUE_START_SIZE 1024
---
> #define QUEUE_START_SIZE 2048
Only in AD5M_Series_Klipper/klippy/extras: bus.pyc
Only in AD5M_Series_Klipper/klippy/extras/display: __pycache__
Only in AD5M_Series_Klipper/klippy/extras: ds18b20.pyc
Only in AD5M_Series_Klipper/klippy/extras: fan.pyc
Only in AD5M_Series_Klipper/klippy/extras: __pycache__
diff -r klipper/klippy/extras/tmc.py AD5M_Series_Klipper/klippy/extras/tmc.py
327,329c327,329
<             did_reset = self.echeck_helper.start_checks()
<             if did_reset:
<                 self.mcu_phase_offset = None
---
>             #did_reset = self.echeck_helper.start_checks()
>             #if did_reset:
>             #    self.mcu_phase_offset = None
diff -r klipper/klippy/extras/virtual_sdcard.py AD5M_Series_Klipper/klippy/extras/virtual_sdcard.py
8c8
< VALID_GCODE_EXTS = ['gcode', 'g', 'gco']
---
> VALID_GCODE_EXTS = ['gcode', 'g', 'gco','gx']
177,178c177,178
<             if fname not in flist:
<                 fname = files_by_lower[fname.lower()]
---
>             #if fname not in flist:
>                 #fname = files_by_lower[fname.lower()]
180c180
<             f = io.open(fname, 'r', newline='')
---
>             f = io.open(fname, 'r', errors='ignore', newline='')
Only in AD5M_Series_Klipper/klippy/kinematics: __pycache__
diff -r klipper/klippy/mcu.py AD5M_Series_Klipper/klippy/mcu.py
744a745,749
>         move_msg = "My Configured MCU '%s' (%d moves)" % (self._name, move_count)
>         logging.info(move_msg)
>         log_info = self._log_info() + "\n" + move_msg
>         self._printer.set_rollover_info(self._name, log_info, log=False)
>
Only in AD5M_Series_Klipper/klippy: mcu.pyc
Only in AD5M_Series_Klipper/klippy: pins.pyc
diff -r klipper/klippy/queuelogger.py AD5M_Series_Klipper/klippy/queuelogger.py
27c27
<             self, filename, when='midnight', backupCount=5)
---
>             self, filename, when='h', interval=4, backupCount=6)
```
So very minor changes and some pyc files and a __pycache__ as a bonus.

**Comment by nativeit** [comment]
@FlashforgeOfficial - Please stop it with the half-measures. Lean into the open source community. Just focus on what you do well--making solid 3D printers at a great price. Keep the source available and buildable. Give us the ability to flash firmware without the OTA crap. 

Let your users **help you** make your hardware better!

**Comment by Lavacrabz** [comment]
@jtenniswood is right @FlashforgeOfficial.

_Release full Klipper support._ You have put together a great piece of Hardware, and your quick change nozzle feature is fantastic. You risk falling behind other companies, like Creality, if you continue to avoid releasing Klipper (ex. the new Phrozen Arco kickstarter is openly advertizing Klipper as a marketing selling point) 

Your 5M series could easily become the Ender 3 of CoreXY machines at the $399/299 discounted price. Think of how many people pick up the Ender as a starting point for price alone on microcenter deals. However, the real thing keeping the Ender selling is the modification community surrounding the Ender 3. If you release Klipper imagine the modification community interest people will have. Please don't miss the opportunity.

If you release Klipper, here are a few points I miss from my Klipper modded Ender 3... you could make a video showing the **new expanded capaiblities of the machines** _with Klipper_ unlocked:

1. Adaptive bed mesh leveling
2. Adaptive purge line before print
3. custom start g-code (ex. heat bed, wait 5min, level bed)
4. timelapse support for other cameras (supporting modification community)
5. PID tuning
6. web-interface (easy to see status from mobile phone)

中文 below
_________________________________________________________________________________________________________________

您已经组装了一台出色的硬件，您的快速更换喷嘴功能非常棒。如果您继续不发布 Klipper（例如，新的 Phrozen Arco 众筹项目公开宣传 Klipper 作为营销卖点），您将面临被其他公司（如 Creality）超越的风险。

您的 5M 系列机器很容易成为 CoreXY 机器中的 Ender 3，售价为 399/299 美元。想象一下，有多少人仅因价格而选择 Ender 作为 Micro Center 交易的起点。然而，真正让 Ender 畅销的是围绕 Ender 3 的修改社区。如果您发布 Klipper，想象一下修改社区的兴趣会有多大。请不要错过这个机会。

如果您发布 Klipper，以下是我从我的 Klipper 改装的 Ender 3 中所遗漏的一些要点… 您可以制作一个视频，展示配备 Klipper 后机器的新扩展功能：

1. 自适应床面网格校准
2. 打印前自适应净化线
3. 自定义启动 G-Code（例如，加热床，等待 5 分钟，校准床面）
4. 支持其他相机的延时摄影（支持修改社区）
5. PID 调整
6. 网页界面（方便通过手机查看状态）

**Comment by mBlinkii** [comment]
I can only agree with the guys here, this printer could become the Ender 3 of the Core XY printer. Creality's success is based on the Ender 3 and that printer was so well received because it was so easy to customise, both the hardware and the software.

Please give us the freedom to customise the software to our liking or at least add a WEB interface that can be accessed from other slicers and allow us to monitor the status of the printer without having FlashPrint open or the PC running all the time. A web interface like Mainsail would be accessible from all possible platforms with a browser (mobile phone, smart TV, tablet, etc...).

**Comment by AJolly** [comment]
This is also missing the ~M601 codes that FF uses to communicate with Flashprint.

**Comment by jvarrone** [comment]
Did you accidentally reply to the wrong thread?Sent from my Galaxy
-------- Original message --------From: AJolly ***@***.***> Date: 4/19/24  7:07 PM  (GMT-05:00) To: FlashforgeOfficial/AD5M_Series_Klipper ***@***.***> Cc: Avpman ***@***.***>, Manual ***@***.***> Subject: Re: [FlashforgeOfficial/AD5M_Series_Klipper] Questions (Issue #1) 
This is also missing the ~M601 codes that FF uses to communicate with Flashprint.

**Comment by AJolly** [comment]
Nope. I'd expect in case they shared their full firmware it it include the
communication macros that they're using to communicate with their software.
But no matter how you look at it what we have is incomplete.

**Comment by jvarrone** [comment]
Thanks. But I don't understand how then people are already using the mod with Orcaslicer?

**Comment by jvarrone** [comment]
Can the Klipper Mod not detect the missing START/END_PRINT codes and perform those steps if those codes are not in the gcode file sent to the modded printer? Seems logical, especially since we can't print without their inclusion.

**Comment by jvarrone** [comment]
This is the response I got from FlashForge support when I recently asked for the source code-

_The modified part is internal information of our company, so we cannot share it with you.
Have a nice day.
Best regards,
Daisy(李文婕) | Aftersales_

**Comment by NomadicCore** [comment]
> This is the response I got from FlashForge support when I recently asked for the source code-
> 
> _The modified part is internal information of our company, so we cannot share it with you. Have a nice day. Best regards, Daisy(李文婕) | Aftersales_

This sounds a whole lot like they aren't actually sharing the full codebase for the firmware used in their printer so they are still violating the klipper licences and missing the whole point of open source. You can't just take open source work, modify it and not share that when the original licences specifically prohibit this.

I'm currently looking at buying a 3d printer and this is massively turning me off Flashforge on pure principle

**Comment by MACCAMENZIES** [comment]
Totally. They are happy to use open source to get their machines running!!

**Comment by rispeed** [comment]
Man i was going to buy one right now but then found it they can't even release the klipper files needed. come on man

**Comment by AJolly** [comment]
Latest response. I'm trying to go back and forth with them and help them to understand what they need to still provide us.

"Based on the open source principle of Klipper, we have made public the part of the Klipper firmware that our company has used. The Klipper firmware on our official website cannot be used directly, please do not upgrade it directly.

The openness of the 3D printing community has driven the development of desktop 3D printing. Flashforge follows the open source protocol, and we open source Klipper firmware here.
Github link: FlashforgeOfficial/AD5M-Series_Klipper (Github. com)

Warning:
Please do not update the Klipper version on your own. Our printer uses Klipper motion control code and has been customized and developed with features suitable for this machine. The public version of Klipper does not directly match the overall firmware.

Modifying Klipper may cause the printer to not work, and there may be some issues that we may not be able to assist you remotely. And after modifying the firmware, the device's warranty will become invalid!

Flashforge has developed Klipper based on a special version and added many new features. The optimized code based on the model structure and hardware can better ensure the stable operation of the device.

If you have your own ideas about Klipper's updates, we do not restrict users from making modifications themselves, but we cannot provide any technical support for this."

**Comment by DangerManiac** [comment]
It's probably useless posting this here as I see zero response from FF but...

**Does this add anything beyond the normal firmware?** **Moonraker?** **Fluidd?** I doubt it but I have to ask, I would specifically like to see Moonraker added. Also yes I know there is a mod to add these but having it official would be better.

One of the biggest things I love about Klipper is the ability to connect to my printers from any of my devices in my house and have a separate KlipperScreen remote set up to control all of my machines.

**What about a PrusaSlicer Profile? How hard can this be for them to do?**

The FlashFPrint 5 slicer sucks, I know it's really dumbed-down slicer for the beginner user but damn. Yes there is an Orca/Flashforge slicer version but then I personally really dislike Orca, it may work great for most people but I hate the interface, just too cluttered for me. Yes I can make my own profile for PrusaSlicer and am currently doing that but it would be nice not to have to make one from scratch.

FlashForge, you have what seems like a pretty good printer but you guys need to up your game on the firmware end, heck at least add the option to enable Moonraker and have it off by default. Hate to have to search for mods from other users for something that should just already be there.

**Comment by BigToyBox** [comment]
FF really needs to start over with a new software team. I have been in the hobby for 10 years. I see the value of the hardware so I bought one to add to my fleet of printers. It floors me that they actually promote using a clogged extruder procedure to unload/change filament! Until there is an UNLOAD button like every other printer in the world, I cannot recommend this printer to anyone. I am running the klipper mod to get the features I am used to. Again, it's shocking that the printer is shipped with out a button to easily load and unload filament. I can imagine first time 3D printer users going thru the process they teach. Sad really.
